pkg/rhcos/release/release.go

```diff
@@ -0,0 +1,157 @@
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+func pull(ctx context.Context, pullSpec string) (types.ImageSource, types.Image, error) {
+	named, err := reference.ParseNamed(pullSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	reference, err := docker.NewReference(named)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sys := &types.SystemContext{}
+	imageSource, err := reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		if err != nil {
+			imageSource.Close()
+		}
+	}()
+
+	unparsed := image.UnparsedInstance(imageSource, nil)
+	// TODO: signature validation
+
+	img, err := image.FromUnparsedImage(ctx, sys, unparsed)
+	return imageSource, img, err
+}
+
+func imageReferences(ctx context.Context, releaseImagePullSpec string) ([]byte, error) {
+	imageSource, image, err := pull(ctx, releaseImagePullSpec)
+	if err != nil {
+		return nil, err
+	}
+	defer imageSource.Close()
+
+	layers := image.LayerInfos()
+	for i := len(layers) - 1; i >= 0; i-- {
+		layer := layers[i]
+		blob, _, err := imageSource.GetBlob(ctx, layer, blobinfocache.NoCache)
+		if err != nil {
+			return nil, err
+		}
+		defer blob.Close()
+
+		// fmt.Println(layer.MediaType) // TODO: blank from quay?  Just assume this is a gzipped tar?
+		gzipReader, err := gzip.NewReader(blob)
+		if err != nil {
+			return nil, err
+		}
+
+		tarReader := tar.NewReader(gzipReader)
+		for {
+			hdr, err := tarReader.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return nil, err
+			}
+
+			if hdr.Name == "release-manifests/image-references" {
+				return ioutil.ReadAll(tarReader)
+			}
+		}
+	}
+
+	return nil, errors.Errorf("no image reference found in %s", transports.ImageName(image.Reference()))
+}
+
+func osContentPullSpec(imageReferences []byte) (string, error) {
+	scheme := runtime.NewScheme()
+	imagev1.Install(scheme)
+	decoder := serializer.NewCodecFactory(scheme).UniversalDecoder(
+		imagev1.GroupVersion,
+	)
+	obj, _, err := decoder.Decode(imageReferences, nil, nil)
+	if err != nil {
+		return "", err
+	}
+
+	imageStream, ok := obj.(*imagev1.ImageStream)
+	if !ok {
+		return "", errors.Errorf("image references is a %q, not an image stream", obj.GetObjectKind())
+	}
+
+	for _, tag := range imageStream.Spec.Tags {
+		if tag.Name == "machine-os-content" {
+			if tag.From == nil {
+				return "", errors.New("machine-os-content tag has an empty 'from' property")
+			}
+
+			if tag.From.Kind != "DockerImage" {
+				return "", errors.Errorf("unrecognized machine-os-content tag from kind %q", tag.From.Kind)
+			}
+
+			return tag.From.Name, nil
+		}
+	}
+
+	return "", errors.New("no machine-os-content tag found")
+}
+
+// RHCOSBuild extracts the RHCOS build version from an OpenShift
+// release image.
+func RHCOSBuild(ctx context.Context, releaseImagePullSpec string) (string, error) {
```

Comment:
can you move this to the top, this is imp func for this package...

Reply:
> can you move this to the top, this is imp func for this package...

Done as part of 327976e5e -> 7c1ad4f6d.